crates/ra_hir/src/ty/infer.rs

```diff
@@ -423,9 +424,21 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                     let typable: Option<TypableDef> = def.into();
                     let typable = typable?;
 
-                    let substs =
-                        Ty::substs_from_path_segment(self.db, &self.resolver, segment, typable);
-                    self.db.type_for_def(typable, Namespace::Types).subst(&substs)
+                    let mut ty = self.db.type_for_def(typable, Namespace::Types);
+
+                    if remaining_index > 0 {
```

Comment:
it's not actually possible for `remaining_index` to be 0 here, because in that case there would be nothing resolved :) I'd suggest removing the if and maybe adding an `assert!(remaining_index > 0)` instead.